flasc/visualization.py

```diff
@@ -639,12 +640,137 @@ def put_label(i):
 
 
 
+def data_plot(
+    x,
+    y,
+    color="b",
+    label="_nolegend_",
+    x_edges=None,
+    ax=None,
+    show_scatter=True,
+    show_bin_points=True,
+    show_confidence=True,
+    alpha_scatter=0.1,
+):
+    """
+    Plot data to a single axis.  Method
+    has options to include scatter of underlying data, specifiying
+    bin edges, and plotting confidence interval.
 
+    Args:
+        x (np.array): abscissa data.
+        y (np.array): ordinate data.
+        color (str, optional): line color.
+            Defaults to 'b'.
+        label (str, optional): line label used in legend.
+            Defaults to '_nolegend_'.
+        x_edges (np.array, optional): bin edges in x data
+            Defaults to None.
+        ax (:py:class:`matplotlib.pyplot.axes`, optional):
+            axes handle for plotting. Defaults to None.
+        show_scatter (bool, optional): flag to control scatter plot.
+            Defaults to True.
+        show_bin_points (bool, optional): flag to control plot of bins.
+            Defaults to True.
+        show_confidence (bool, optional): flag to control plot of
+            confidence interval. Defaults to True.
+        alpha_scatter (float, optional): Alpha for scatter
+            plot. Defaults to 0.5.
+
+    """
+
+    # Check the length of x equals length of y
+    if len(x) != len(y):
+        raise ValueError("x and y must be the same length")
     
+    # Check that x is not empty
+    if len(x) == 0:
+        raise ValueError("x is empty")
     
 
-    
+    # Declare ax if not provided
+    if ax is None:
+        _, ax = plt.subplots()
+
+    # Put points ino dataframe
+    df = pd.DataFrame({"x": x, "y": y})
 
+    # If x_edges not provided, use 50 bins over range of x
+    if x_edges is None:
+        x_edges = np.linspace(df["x"].min(), df["x"].max(), 50)
 
+    # Define x_labels as bin centers
+    x_labels = (x_edges[1:] + x_edges[:-1]) / 2.0
 
+    # Bin data
+    df["x_bin"] = pd.cut(df["x"], x_edges, labels=x_labels)
+
+    # Get aggregate statistics
+    df_agg = df.groupby("x_bin").agg(
+        {"y": ["count", "std", "min", "max", "mean"]}
+    )
+    # Flatten column names
+    df_agg.columns = ["_".join(c) for c in df_agg.columns]
+
+    # Reset the index
+    df_agg = df_agg.reset_index()
+
+    # Add the confidence interval of the mean to df_agg
+    df_agg["y_confidence"] = 1.96 * df_agg["y_std"] / np.sqrt(df_agg["y_count"])
```

Comment:
Where does this equation come from? Can we add a variable called `confidence_interval` to allow the user to specify the confidence interval?

Reply:
Yes, that's a good idea, I think 1.96 is the value for 95% confidence but the general equation I think we can add a more generic based on this:

http://www.stat.yale.edu/Courses/1997-98/101/confint.htm#:~:text=The%20critical%20value%20for%20a,%3D%20(100.86%2C%20102.78).

or maybe a more canned approach like:

https://stackoverflow.com/questions/15033511/compute-a-confidence-interval-from-sample-data

?

Reply:
Thanks guys!  I found a way to do this generally online (using st but a slightly different implementation) but I think this one is maybe now wrong, it's using a t-distribution, what do you think?  Maybe I should've grabbed your approach @misi9170

Reply:
@paulf81 I think what you have, using the t-distribution, is good (it's a little more correct than the z-score approach, which assumes the data set is large enough that the means can be treated as normal---the t-distribution doesn't make that assumption, and approaches the z-score method as the number of samples becomes large). If rearranged the code a little in my commits, but have not changed the method.